hw2-generics

dzrgln opened the pull request.

Commit:
hw2-generics

FlameXander commented on hw2-generics/src/main/java/ru/otus/hw2/generics/Apple.java:
```diff
@@ -0,0 +1,8 @@
+package ru.otus.hw2.generics;
```
Это кажется третье дз (с учетом добавленного по тестированию)

FlameXander commented on hw2-generics/src/main/java/ru/otus/hw2/generics/Box.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public boolean compare(Box<? extends Fruit> anotherBox) {
+        return weight() == anotherBox.weight();
```
Числа с плавающей запятой не должны сравниваться через ==
Что будет если anotherBox == null?

FlameXander commented on hw2-generics/src/main/java/ru/otus/hw2/generics/Box.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public void pour(Box<T> anotherBox) {
+        fruits.addAll(anotherBox.fruits);
```
Что будет если anotherBox == null или anotherBox == this?

Commit:
hw3-generics

1. Пакет переименован в hw3
2. Добавлена валидация коробок

Commit:
hw3-generics

Оптимизация импортов

FlameXander commented on hw2-generics/src/main/java/ru/otus/hw3/generics/Box.java:
```diff
@@ -0,0 +1,43 @@
+
+    public void pour(@NotNull Box<T> anotherBox) {
+        if (anotherBox.equals(this)) {
+            throw new IllegalArgumentException();
```
Лучше лишний раз исключения не бросать, можно просто return сделать. Если из коробки в себя же фрукты пересыпали, то они там и остались

dzrgln replied:
Поправил. 
Но не могли бы вы пояснить, почему в этом случае лучше без исключения? 
Я предполагал, что пересыпка коробки из самой в себя действие странное и не имеющее смысла. В моей голове, что если пользователь такое делает, то делает по ошибке. И лучшим вариантом будет быстро упасть с исключением.

FlameXander replied:
Если указать в качестве коробки для пересыпания саму себя, то мне кажется что это действие равносильно тому что не надо ничего делать. С другой стороны можно бросить исключение что такое действие подозрительно. Лично мне кажется что в некритичных случаях проще пропустить действие, чем бросать исключения, чтобы на ровном месте не ломать программу, и чем решать за других разработчиков что в их алгоритмах они допустили какие-то логические ошибки (это уж пусть они на тестах выявляют). Но на самом деле оба варианта имеют место быть, на такой ограниченной задаче нельзя 100% сказать что один из вариантов явно выигрышный

FlameXander commented on settings.gradle:
```diff
@@ -1,3 +1,4 @@
 rootProject.name = 'ru.otus.java.pro'
 include 'hw1-gradle'
+include 'hw2-generics'
```
Пакет переименовали, а модуль?)

Commit:
hw3-generics

1. Переименование модуля
2. Изменение логики в случае пересыпания из коробки саму в себя
